pkg/controller/worker/genericactuator/actuator_resore.go

```diff
@@ -0,0 +1,172 @@
+// Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package genericactuator
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	extensionscontroller "github.com/gardener/gardener-extensions/pkg/controller"
+	workercontroller "github.com/gardener/gardener-extensions/pkg/controller/worker"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	machinev1alpha1 "github.com/gardener/machine-controller-manager/pkg/apis/machine/v1alpha1"
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (a *genericActuator) Restore(ctx context.Context, worker *extensionsv1alpha1.Worker, cluster *extensionscontroller.Cluster) error {
+	workerDelegate, err := a.delegateFactory.WorkerDelegate(ctx, worker, cluster)
+	if err != nil {
+		return errors.Wrapf(err, "could not instantiate actuator context")
+	}
+
+	// Generate the desired machine deployments.
+	wantedMachineDeployments, err := workerDelegate.GenerateMachineDeployments(ctx)
+	if err != nil {
+		return errors.Wrapf(err, "failed to generate the machine deployments")
+	}
+
+	// Get the list of all existing machine deployments.
+	existingMachineDeployments := &machinev1alpha1.MachineDeploymentList{}
+	if err := a.client.List(ctx, existingMachineDeployments, client.InNamespace(worker.Namespace)); err != nil {
+		return err
+	}
+
+	// Parse the worker state to a separete machineDeployment states and attach them to
+	// the corresponding machineDeployments which are to be deployed later
+	a.logger.Info("Extracting the worker status.state", "worker", fmt.Sprintf("%s/%s", worker.Namespace, worker.Name))
+	if err := a.addStateToMachineDeployment(ctx, worker, wantedMachineDeployments); err != nil {
+		return err
+	}
+
+	wantedMachineDeployments = removeWantedDeploymentWithoutState(wantedMachineDeployments)
+
+	// Delete the machine-controller-manager. During restoration MCM must not exists
```

Comment:
@hardikdr @prashanth26 @amshuman-kr can you say if this step is necessary if we perform the steps in the following order -> create machines -> create machineset -> create machinedeployment.
At this point we still do not have machineclasses so it's most likely necessary.